sky/data/data_utils.py

```diff
@@ -126,6 +155,149 @@ def verify_gcs_bucket(name: str) -> bool:
+
+    Returns:
+        Name of the resource group the given storage account belongs to.
+    """
+    if storage_client is None:
+        storage_client = create_az_client('storage')
+    for account in storage_client.storage_accounts.list():
+        if account.name == storage_account_name:
+            # Extract the resource group name from the account ID
+            # An example of account.id would be the following:
+            # /subscriptions/{subscription_id}/resourceGroups/{resource_group_name}/providers/Microsoft.Storage/storageAccounts/{container_name} # pylint: disable=line-too-long
+            split_account_id = account.id.split('/')
+            assert len(split_account_id) == 9
+            resource_group_name = split_account_id[4]
+            return resource_group_name
+    # resource group cannot be found when using container not created
+    # under the user's subscription id, i.e. public container or
+    # private containers not belonging to the user.
+    return None
+
+
+def get_az_storage_account_key(
+    storage_account_name: str,
+    resource_group_name: Optional[str] = None,
+    storage_client: Optional[Client] = None,
+    resource_client: Optional[Client] = None,
+) -> Optional[str]:
+    """Returns access key of the given name of storage account.
+
+    Args:
+        storage_account_name: str; Name of the storage account
+        resource_group_name: str; Name of the resource group the passed storage
+            account belongs to.
+        storage_clent: Optional[Client]; client object facing Storage
+        resource_client: Optional[Client]; client object facing Resource
+
+    Returns:
+        One of the few access keys to the given storage account
+    """
+    # resource_group_name is None when using a public container or
+    # a private containers not belonging to the user.
+    if resource_group_name is None:
+        return None
+    if resource_client is None:
+        resource_client = create_az_client('resource')
+    if storage_client is None:
+        storage_client = create_az_client('storage')
+    attempt = 0
+    backoff = common_utils.Backoff()
+    while True:
+        try:
+            resources = resource_client.resources.list_by_resource_group(
+                resource_group_name)
+            # resource group is either created or read when Storage initializes.
+            assert resources is not None
+            for resource in resources:
+                if (resource.type == 'Microsoft.Storage/storageAccounts' and
+                        resource.name == storage_account_name):
+                    storage_account_keys = storage_client.storage_accounts.list_keys(
+                        resource_group_name, storage_account_name)
+                    storage_account_keys = [
+                        key.value for key in storage_account_keys.keys
+                    ]
+            storage_account_key = storage_account_keys[0]
+            return storage_account_key
```

Comment:
```suggestion
        try:
            storage_account_keys = None
            resources = resource_client.resources.list_by_resource_group(
                resource_group_name)
            # resource group is either created or read when Storage initializes.
            assert resources is not None
            for resource in resources:
                if (resource.type == 'Microsoft.Storage/storageAccounts' and
                        resource.name == storage_account_name):
                    storage_account_keys = storage_client.storage_accounts.list_keys(
                        resource_group_name, storage_account_name)
                    assert storage_account_keys is None, "Multiple storage account named {storage_account_name} found in resources group {resource_group_name}."
                    storage_account_keys = [
                        key.value for key in storage_account_keys.keys
                    ]
            assert storage_account_keys is not None, f"Storage account {storage_account_name} not found in resources group {resource_group_name}."
            assert len(storage_account_keys) == 1, "some debug info here"
            storage_account_key = storage_account_keys[0]
            return storage_account_key
```
Try to avoid the `UnboundedLocalError`, and multiple assignment to `storage_account_keys`. Also shows more information in the assertion statement.

Reply:
The `UnboundLocalError` occurs when the creation of the storage account did not propagate to Azure yet. Waiting for couple of seconds allows for the creation to propagate and let retreive the storage account key from tha API call.

Reply:
I mean causing an unbound local error seems a little bit messy and hacky to me. This should not happen for a robust code. We should, for example, have a function to return an optional string to indicate the storage account name, and `None` means not found yet instead. Or possibly to raise a customized exception and catch it. Open to other possible solutions as well

Reply:
Great point. Resolved at [01b48a4](https://github.com/skypilot-org/skypilot/pull/3032/commits/01b48a4830dd8ac97c3dc1c1c4334e15bdd6b7f4), [2bd5ec7](https://github.com/skypilot-org/skypilot/pull/3032/commits/2bd5ec773780a3995e0d3990eb7db5492088e291).
Try except block to catch `UnboundLocalError` was replaced with an `if` statement to check if `storage_account_keys` is `None`.